syntax = 'proto3';

package magma.broker;

import "broker/protos/common.proto";
//option go_package = "magma/broker/cloud/go/protos";

service Brokerd {
    // -- auth protocol
    // modified Authentication-Information (Code 318, S6a)
    rpc BrokerAuthenticationInformation (BrokerAuthenticationInformationRequest) returns (BrokerAuthenticationInformationAnswer) {}

    // ..update location etc.

    // -- billing protocol
    rpc ReportUsage(Usage) returns (Void) {
    }
    // -- handover protocol
}

message BrokerAuthenticationInformationRequest {
    AuthenticationInformationRequest content = 1;
    string signature = 2;
}

message BrokerAuthenticationInformationAnswer {
    AuthenticationInformationAnswer content = 1;
    string signature = 2;
}

// Authentication Information Request (Section 7.2.5, S6a)
message AuthenticationInformationRequest {
    // Subscriber identifier
    string user_name = 1;
    // Visted site identifier
    bytes visited_plmn = 2;
    // Number of vectors to request in response
    uint32 num_requested_eutran_vectors = 3;
    // Indicates to the HSS the values are requested for immediate attach
    bool immediate_response_preferred = 4;
    // Concatenation of RAND and AUTS in the case of a resync attach case
    bytes resync_info = 5;
}


// Authentication Information Answer (Section 7.2.6, S6a)
message AuthenticationInformationAnswer {
    // EPC error code on failure
    ErrorCode error_code = 1;
    // Authentication vectors matching the requested number
    repeated EUTRANVector eutran_vectors = 2;

    // For details about fields read 3GPP 33.401
    message EUTRANVector {
        bytes rand = 1;
        bytes xres = 2;
        bytes autn = 3;
        bytes kasme = 4;
    }
}

// ErrorCode reflects Experimental-Result values which are 3GPP failures
// to be processed by EPC. Diameter Base Protocol errors are reflected in gRPC status code
enum ErrorCode {
    UNDEFINED                = 0;
    // Default success code
    MULTI_ROUND_AUTH         = 1001;
    SUCCESS                  = 2001;
    LIMITED_SUCCESS          = 2002;
    COMMAND_UNSUPORTED       = 3001;
    UNABLE_TO_DELIVER        = 3002;
    REALM_NOT_SERVED         = 3003;
    TOO_BUSY                 = 3004;
    LOOP_DETECTED            = 3005;
    REDIRECT_INDICATION      = 3006;
    APPLICATION_UNSUPPORTED  = 3007;
    INVALIDH_DR_BITS         = 3008;
    INVALID_AVP_BITS         = 3009;
    UNKNOWN_PEER             = 3010;
    AUTHENTICATION_REJECTED  = 4001;
    OUT_OF_SPACE             = 4002;
    ELECTION_LOST            = 4003;
    AUTHORIZATION_REJECTED   = 5003;
    // Permanent Failures 7.4.3
    USER_UNKNOWN             = 5001;
    UNKNOWN_SESSION_ID       = 5002;
    UNKNOWN_EPS_SUBSCRIPTION = 5420;
    RAT_NOT_ALLOWED          = 5421;
    ROAMING_NOT_ALLOWED      = 5004;
    EQUIPMENT_UNKNOWN        = 5422;
    UNKOWN_SERVING_NODE      = 5423;
    // Transient Failures 7.4.4
    AUTHENTICATION_DATA_UNAVAILABLE = 4181;
}
